compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfigAccess.java

```diff
@@ -146,6 +149,19 @@ public static boolean jvmciGE(Version v) {
         JVMCI_VERSION = Version.parse(vmVersion);
         JVMCI_PRERELEASE = vmVersion.contains("SNAPSHOT") || vmVersion.contains("internal") || vmVersion.contains("-dev");
         JVMCI = JVMCI_VERSION != null || JVMCI_PRERELEASE;
+        int jdkUpdateVers = UNSET;
```

Comment:
You should use a regular expression to parse the value of the `java.vm.version` system property. Here are some sample values:
11.0.7+8-LTS-jvmci-20.1-b02
25.241-b07-jvmci-20.1-b01

Reply:
`UNSET` is just the default. I was trying to avoid all this version parsing mess. In particular, for this issue I'm not sure it's worth doing the right thing for JDK 8, is it? If that's needed, I'd rather have a specialized class which only deals with that and is unit-tested.

So if `Runtime.version().update()` cannot be used, how about `Runtime.Version.parse(<value>)` where `<value>` could be coming from `java.vm.version` property as you say.

Reply:
So I'd be something like: http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/e6178063e9dd/test/sun/misc/Version/Version.java#l139 for `java.vm.version` parsing for JDK 8. For JDK 9+ we really should leverage the JDK for the parsing rather than writing our own.

Reply:
For 8, *only* JVMCI JDKs can be used (currently) so we can rely on JVMCI versions there.

Reply:
I think it's simplest to just use regular expressions. Something like this:
```
diff --git a/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfig.java b/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfig.java
index 45eeb0aff99..4d02f306304 100644
--- a/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfig.java
+++ b/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfig.java
@@ -154,7 +154,7 @@ public class GraalHotSpotVMConfig extends GraalHotSpotVMConfigAccess {
     private final boolean useSquareToLenIntrinsic = getFlag("UseSquareToLenIntrinsic", Boolean.class, false, isJDK8OrJDK11Plus);
     public final boolean useVectorizedMismatchIntrinsic = getFlag("UseVectorizedMismatchIntrinsic", Boolean.class, false, isJDK11Plus);
     public final boolean useFMAIntrinsics = getFlag("UseFMA", Boolean.class, false, JDK >= 9);
-    public final int useAVX3Threshold = getFlag("AVX3Threshold", Integer.class, 4096, osArch.equals("amd64") && (JVMCI ? JDK >= 11 : JDK >= 14));
+    public final int useAVX3Threshold = getFlag("AVX3Threshold", Integer.class, 4096, osArch.equals("amd64") && (JVMCI ? JDK >= 11 : JDK == 11 ? JDK_UPDATE >= 6 : JDK >= 14));
 
     public final boolean preserveFramePointer = getFlag("PreserveFramePointer", Boolean.class);
 
diff --git a/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfigAccess.java b/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfigAccess.java
index b99cbd6975b..b9f84d97132 100644
--- a/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfigAccess.java
+++ b/compiler/src/org.graalvm.compiler.hotspot/src/org/graalvm/compiler/hotspot/GraalHotSpotVMConfigAccess.java
@@ -30,6 +30,8 @@ import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 import org.graalvm.compiler.debug.Assertions;
@@ -137,6 +139,7 @@ public class GraalHotSpotVMConfigAccess {
     }
 
     public static final int JDK = JavaVersionUtil.JAVA_SPEC;
+    public static final int JDK_UPDATE;
     public static final boolean IS_OPENJDK = getProperty("java.vm.name", "").startsWith("OpenJDK");
     public static final Version JVMCI_VERSION;
     public static final boolean JVMCI;
@@ -146,6 +149,13 @@ public class GraalHotSpotVMConfigAccess {
         JVMCI_VERSION = Version.parse(vmVersion);
         JVMCI_PRERELEASE = vmVersion.contains("SNAPSHOT") || vmVersion.contains("internal") || vmVersion.contains("-dev");
         JVMCI = JVMCI_VERSION != null || JVMCI_PRERELEASE;
+
+        String javaVersion = getProperty("java.vm.version");
+        Matcher m = Pattern.compile("(?:1.8.0_(\\d+))|(?:(\\d+)\\.(\\d+)\\.(\\d+))").matcher(javaVersion);
+        if (!m.matches()) {
+            throw new JVMCIError("Unexpected java.version value: " + javaVersion);
+        }
+        JDK_UPDATE = Integer.parseInt(m.group(1) != null ? m.group(1) : m.group(4));
     }
 
     private final List<String> missing = new ArrayList<>();
```

Reply:
That's what I mean. It'll fail for all sorts of weird versions. I'm pretty sure this fails for my hypothetical JDK 8 here: `java.vm.version = 25.242-b08`. Since we don't need to cover JDK 8 - covered by jvmci, why would we want to craft a custom parser which then fails in strange ways. Example for JDK 14 I have here: `java.vm.version = 14+36`.

Reply:
Can there really be a lot of variation in values for the `java.version` system property? The specification for [Runtime.Version](https://download.java.net/java/GA/jdk14/docs/api/java.base/java/lang/Runtime.Version.html) includes a regular expression for a _version number_ and I think `java.version` will always be a version number as opposed a more variable _version string_. In any case a version string is specified to always start with a version number.

If your really want to rely on using `Runtime.Version` for parsing `java.version` (which I'm not objecting to), then you will have to use an [overlay](https://github.com/graalvm/mx/blob/master/README.md#versioning-sources-for-different-jdk-releases) which is a larger change.